### PR TITLE
Feature/all by type

### PR DESCRIPTION
**Pull request by khatton-ft**
- Support get all concepts by type, if the concept type is mapped to a type for the ElasticSearch index. Presently only http://www.ft.com/ontology/Genre is mapped.
- Update to ElasticSearch version 5.
- Includes tests that run against a local ES. These can be skipped using `-short`.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -54,26 +57,26 @@ func main() {
...
-		esRegion:   *esRegion,
-	}
+	accessConfig := service.NewAccessConfig(*accessKey, *secretKey, *esEndpoint)
+
+	log.SetLevel(log.InfoLevel)
 
 	app.Action = func() {
 		logStartupConfig(port, esEndpoint, esRegion, esIndex, searchResultLimit)
-		client, err := newElasticClient(accessConfig.accessKey, accessConfig.secretKey, &accessConfig.esEndpoint, &accessConfig.esRegion)
+		esClient, err := service.NewElasticClient(*esRegion, accessConfig /*accessConfig.accessKey, accessConfig.secretKey, &accessConfig.esEndpoint, &accessConfig.esRegion*/)
```
Comment (izzyblues) [inline main.go]: Any reason for this commented code?

File: service/search_test.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+	esURL := os.Getenv("ELASTICSEARCH_TEST_URL")
+	if strings.TrimSpace(esURL) == "" {
+		t.Fatal("Please set the environment variable ELASTICSEARCH_TEST_URL to run ElasticSearch integration tests (e.g. export ELASTICSEARCH_TEST_URL=http://localhost:9200). Alternatively, run `go test -short` to skip them.")
+	}
+
+	return esURL
+}
+
+func writeTestConcepts(ec *elastic.Client) []string {
```
Comment (peteclark-ft) [inline service/search_test.go]: Should we clean these up?

File: main.go (hunk 1 above)
Comment (khatton-ft) [inline main.go]: PEBKAC, fixed :)

File: resources/handler_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
+package resources
```
Comment (khatton-ft) [inline resources/handler_test.go]: :+1: Added further tests for no `type`, blank `type`, many `types` and a non-empty `q`. I've been quite strict about what we return, as a first draft, so these cases generally return HTTP 400 responses.

File: service/es_client.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+type AWSSigningTransport struct {
+	HTTPClient  *http.Client
+	Credentials awsauth.Credentials
+}
+
+// RoundTrip implementation
+func (a AWSSigningTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	return a.HTTPClient.Do(awsauth.Sign4(req, a.Credentials))
```
Comment (khatton-ft) [inline service/es_client.go]: :+1: ~Plagiarised~Did some research and adapted some code from the Go team's OAuth library which does just that.

File: service/search.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+			esConcept := EsConceptModel{}
+			json.NewDecoder(bytes.NewReader(by)).Decode(&esConcept)
+
+			concept := ConvertToSimpleConcept(esConcept, c.Type)
+			concepts = append(concepts, concept)
+		}
+	}
+	sort.Sort(concepts)
+
+	return concepts, nil
```
Comment (khatton-ft) [inline service/search.go]: Absent-mindedness :smile:. Fixed.

File: resources/handler.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+func NewHandler(service service.ConceptSearchService) *Handler {
+	return &Handler{service}
+}
+
+func (h *Handler) ConceptSearch(w http.ResponseWriter, req *http.Request) {
+	query := req.URL.Query()
+
+	values := query["q"]
```
Comment (khatton-ft) [inline resources/handler.go]: As of this branch, it serves only to check that there is no value of `q`, to avoid false expectations on the part of the client. I couldn't see a way to route based on query string parameters that would also work with Vestigo (with a hunch that we might move in that direction); plus the rules about how to route based on query string parameters aren't trivial, so I think we'll end up implementing that in code here anyway.

File: resources/handler.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	} else {
+		err = service.ErrInvalidConceptType
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+	if err != nil {
+		response["message"] = err.Error()
+
+		if err == service.ErrInvalidConceptType {
+			w.WriteHeader(http.StatusBadRequest)
```
Comment (khatton-ft) [inline resources/handler.go]: I wonder ... do you think requesting `?type=unknown-type` should be a `404` rather than a `400`?

File: service/model_test.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+func TestEsType(t *testing.T) {
+	assert.Equal(t, "genres", esType("http://www.ft.com/ontology/Genre"), "known type conversion")
+	assert.Equal(t, "", esType("http://www.ft.com/ontology/Foo"), "unknown type conversion")
+}
+
+func TestFtType(t *testing.T) {
+	assert.Equal(t, "http://www.ft.com/ontology/Genre", ftType("genres"), "known type conversion")
+	assert.Equal(t, "", ftType("tardigrades"), "unknown type conversion")
```
Comment (khatton-ft) [inline service/model_test.go]: :laughing: As Lord Reith said, to educate, inform and entertain ...

File: service/search_test.go (hunk 2 above)
Comment (khatton-ft) [inline service/search_test.go]: Fair point, I'll look into that. As the test code doesn't use the signing transport, there's no way it could write to an AWS ES instance, and this works with an empty ES out-of-the-box (for CircleCI).

File: circle.yml [hunk 9]
```diff
@@ -1,29 +1,41 @@
 machine:
```
Comment (khatton-ft) [inline circle.yml]: I've had quite enough Ordeal-By-Circle for now; I think that would be scope creep for this PR.

File: service/search_test.go (hunk 2 above)
Comment (khatton-ft) [inline service/search_test.go]: :+1: See https://github.com/Financial-Times/concept-search-api/pull/20/commits/6f16a15df3fab073942234b3d4a70449a6764aac - I've used the `suite` package to get setup/teardown functionality.

Comment (peteclark-ft) [inline service/search_test.go]: Nice! Not used that before

File: resources/handler.go (hunk 7 above)
Comment (peteclark-ft) [inline resources/handler.go]: I'm not sure... I think for this PR we can park this and leave it as a `400`, but maybe we should consider checking if it's roughly in the expected format of a Concept in a future change? 

i.e. something to validate that it's in the format `http://www.ft.com/ontology/*`, if it is, then `404` if not `400`
